### PR TITLE
Add initial implementation of Imviz

### LINKED ISSUE #513: JDAT-1218: Pan/Zoom
Pan/Zoom using mouse click-and-drag.

1. Clicking should do something by default (esp. middle click)
2. Need to decide on details of click and drag behavior

https://jira.stsci.edu/browse/JDAT-1218

### LINKED ISSUE #515: JDAT-1219: Stretch/Colormaps
* Interactive Stretch adjust.
* A "reasonable number" of scale options.

https://jira.stsci.edu/browse/JDAT-1219

### LINKED ISSUE #517: JDAT-1225: Image Overlays and Transparency
Support transparent overlays of multiple images.

https://jira.stsci.edu/browse/JDAT-1225

### LINKED ISSUE #522: JDAT-1226: Export region selections as astropy regions
Be able to export selected regions as Astropy `regions` to other tools (plotting, photometry measurements, etc.)

https://jira.stsci.edu/browse/JDAT-1226

### LINKED ISSUE #523: JDAT-1393: Create app version of Imviz
Create a standalone version of Imviz that runs "outside" of a notebook. Should include as much v1.0 required functionality as possible, excluding notebook API and interactions.

https://jira.stsci.edu/browse/JDAT-1393

### PR DESCRIPTION
**Pull request by astrofrog**
This is an experimental PR for now and a work in progress, which aims to implement a very simple imviz viewer (one with one viewer and one with two).

EDIT 2 April 2021: Note that this now requires the latest stable release of bqplot-image-gl as well as this branch of glue-jupyter: https://github.com/glue-viz/glue-jupyter/pull/213. Also I've now added a notebook in this PR to demonstrate the features.

Main to-dos:

* [x] Figure out how to automatically link the WCSes on data loading - there is infrastructure in glue to do this so it should be a pretty simple call, I can investigate this.
* [x] Figure out how to avoid having two configurations, one for one-panel imviz and one for two-panel imviz

(I'll update this list as other things come up)

# Ticket tracking

- [x] Finish syncing.
- [x] Does @astrofrog agree with this?

Fix #513 , fix #515 , fix #517 , fix #522 , fix #523

Partially address #514 (shows the info but not in the desired way), #518, #519, #520

Close https://jira.stsci.edu/browse/JDAT-1192

### CODE REVIEW

**Comment by pllim** [comment]
I'll leave my feedback from your first video demo here, for future reference. It was a great demo. Thanks!

* Would be nice to have all the commands update the existing window instead of creating new displays, but I understand this is a quick "proof of concept"
* For the region, does it support "markers" like "x", "+", or "*" on click? Does region support only cover 2D shapes?
* I think we want cursor info outside of the image... though maybe this will be naturally solved if we decide to go down the astrowidgets API via abstract class route...

**Comment by pllim** [comment]
Note to self: https://gist.github.com/astrofrog/7b8ebb3677ff1f5877c6ae82fc8876ec

**Comment by astrofrog** [comment]
I've now tidied things up a lot following improvements in bqplot-image-gl and glue-jupyter and made the various pieces of functionality into tools or baked in to the imviz viewer directly. The example notebook is included here. Note that this now requires the latest stable release of bqplot-image-gl as well as this branch of glue-jupyter: glue-viz/glue-jupyter#213

@pllim:

> Would be nice to have all the commands update the existing window instead of creating new displays, but I understand this is a quick "proof of concept"

Yes I guess the difficulty is that if the notebook is long enough one would need to keep scrolling between reading text and the single viewer, so this keeps each example next to the viewers. Also as it is once all the cells are run you can go and run the different examples whereas with a single viewer you can't do that. I think it's just a matter of preference - if you want to reorganize the notebook to have only one viewer, that's ok too.

> For the region, does it support "markers" like "x", "+", or "*" on click?

No but it would be easy to do so I think - for now this section was meant more as a 'joke' in the sense that DS9 used to just add lots of green circles wherever you clicked even if you didn't want it. But we could easily keep track of the regions and have the coordinates accessible to the user, and also have custom markers.

> Does region support only cover 2D shapes?

For now yes - at least in the imviz viewer. What use cases are you thinking of?

> I think we want cursor info outside of the image... though maybe this will be naturally solved if we decide to go down the astrowidgets API via abstract class route... (astropy/astrowidgets#126)

We can tweak things like this if needed - I'm not sure if astrowidgets would solve this now that we are just going to think of it as an API wrapper. Here I think instead we might want to change the figure margins and move the label down a bit so it falls in the bottom margin.

**Comment by astrofrog** [comment]
I've added a running 'todo' list to the description of this PR.

**Comment by astrofrog** [comment]
New demo video based on this PR: https://www.dropbox.com/s/8zc3iox33s5pw7y/imviz_demo_20200408.mov?dl=0

A couple of quick notes:

* There is a bug with the opacity - when the opacity slider is at max, only one layer should be visible (I'll investigate but won't have time to update the video, this is just a disclaimer in case anyone asks)
* I don't actually expect users to have to do the manoeuvre with changing the reference data in the second viewer etc - we can make this easier and automatic, but this needs more discussion for now as to how we would like it to behave.

File: notebooks/ImvizExample.ipynb [hunk 1]
```diff
@@ -0,0 +1,228 @@
...
+   "metadata": {
+    "scrolled": false
+   },
+   "outputs": [],
+   "source": [
+    "from jdaviz import ImViz\n",
+    "import matplotlib.pyplot as plt\n",
+    "\n",
+    "imviz = ImViz()\n",
+    "imviz.load_data('jw01072001001_01101_00001_nrcb1_cal.fits')\n",
```
Comment (pllim) [inline notebooks/ImvizExample.ipynb:51]: Can we replace this with something anyone have access to? Like something in `astropy-data` or STScI Box URL thingy?

**Review by pllim — COMMENTED** [review]
Comments from video demo:

* I still want the cursor info outside of image display for several reasons:
  * It is distracting to have the info cover parts of real data.
  * When you change colormaps, the cursor info can be hard to see.
  * In the future, it will be easier to have user show/hide cursor info or change its positioning w.r.t. image display, if desired.
  * In the future, it is also easier to implement displaying multiple WCS systems if it does not obstruct data.
* Is X and Y in cursor info 0-indexed or 1-indexed? Is there an option to allow user to specify this?
* I think there can be a simple hack-fix for opacity bug: Special handling for opacity at 0 and 1.
* Where will the keyboard magic be documented? For instance, I wouldn't know pressing "b" will blink unless I watch that video.
* I see in the JIRA tickets that there is a desire to have multiple viewers as tabs, instead of side by side. Future work, but will `b` be able to blink tabs too? Something to think about but out of scope here.
* Not clear from the demo: Will zoom also match the WCSes?

**Comment by pllim** [comment]
p.s. Please get astrofrog/jdaviz#1 in here before merge.

**Comment by pllim** [comment]
p.p.s. Should we pin `bqplot-image-gl>=1.3.0` in `setup.cfg` or is it too early to do so?

**Comment by pllim** [comment]
> Figure out how to automatically link the WCSes on data loading - there is infrastructure in glue to do this so it should be a pretty simple call, I can investigate this.

I am not sure if this is desirable. If I am displaying calibration data (e.g., bias, dark), I wouldn't want this to be automatic. I think it should not be on by default.

> Figure out if we want to have the 'click to draw shapes' functionality demonstrated at the end of the notebook as a way of building up a table of targets for example, distinct from the ability to draw 2d selection regions.

Yes to building a table of targets, but I never seen a need to do this with 2D regions. A common use case is grabbing only the X and Y on clicks for photometry or catalog matching. So, I would prioritize "click to show points" over "click to draw shapes."

**Comment by pllim** [comment]
Re: https://github.com/spacetelescope/jdaviz/pull/429#issuecomment-815919232

I was told by POs that 1D shape isn't a priority right now, so you can ignore me there and stick to 2D regions.

**Comment by pllim** [comment]
Also see astrofrog/jdaviz#2

**Comment by pllim** [comment]
Note to self: This blocks https://jira.stsci.edu/browse/JDAT-1406

File: notebooks/ImvizExample.ipynb (hunk 1 above)
Comment (astrofrog) [reply notebooks/ImvizExample.ipynb]: Done, for now just using data from data.astropy.org and later we could switch to JWST simulated (or actual eventually!) data

**Comment by astrofrog** [comment]
I've now updated this PR so that the auto-linking happens when the 'match WCS' button is clicked which I think is reasonable. So I think in the interest of moving forward, if the CI passes this can probably be merged and we can open follow-up issues and PRs for your comments @pllim?

**Comment by pllim** [comment]
@astrofrog , I opened some follow-up issues and assigned to you those you explicitly said you were going to investigate. FYI.

**Pull request merged**